src/rustzcash.rs

```diff
@@ -1563,3 +1568,86 @@ pub extern "system" fn librustzcash_sapling_proving_ctx_init() -> *mut SaplingPr
 pub extern "system" fn librustzcash_sapling_proving_ctx_free(ctx: *mut SaplingProvingContext) {
     drop(unsafe { Box::from_raw(ctx) });
 }
+
+#[no_mangle]
+pub extern "system" fn librustzcash_zip32_xsk_master(
+    seed: *const c_uchar,
+    seedlen: size_t,
+    xsk_master: *mut [c_uchar; 169],
+) {
+    let seed = unsafe { std::slice::from_raw_parts(seed, seedlen) };
+
+    let xsk = zip32::ExtendedSpendingKey::master(seed);
+
+    xsk.write(&mut (unsafe { &mut *xsk_master })[..])
+        .expect("should be able to serialize an ExtendedSpendingKey");
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_zip32_xsk_derive(
+    xsk_parent: *const [c_uchar; 169],
+    i: uint32_t,
+    xsk_i: *mut [c_uchar; 169],
+) {
+    let xsk_parent = zip32::ExtendedSpendingKey::read(&unsafe { *xsk_parent }[..])
+        .expect("valid ExtendedSpendingKey");
+    let i = zip32::ChildIndex::from_index(i);
+
+    let xsk = xsk_parent.derive_child(i);
+
+    xsk.write(&mut (unsafe { &mut *xsk_i })[..])
+        .expect("should be able to serialize an ExtendedSpendingKey");
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_zip32_xfvk_derive(
+    xfvk_parent: *const [c_uchar; 169],
+    i: uint32_t,
+    xfvk_i: *mut [c_uchar; 169],
+) -> bool {
+    let xfvk_parent = zip32::ExtendedFullViewingKey::read(&unsafe { *xfvk_parent }[..])
+        .expect("valid ExtendedFullViewingKey");
+    let i = zip32::ChildIndex::from_index(i);
+
+    let xfvk = match xfvk_parent.derive_child(i) {
```

Comment:
Add somewhere descriptive error message for hardened child index?
The one in `derive_child` just returns an error without a message.

Reply:
Where should that happen exactly? Isn't this librustzcash?

Reply:
Sorry, I meant in the refactor work happening in the master branch. This PR is merging into a 2.0.1-specific branch, which will later be merged into master.